### ISSUE TITLE
2.x: Completable.takeUntil() behavior suggestion

### ISSUE THREAD

**Issue by skgmn** [ReactiveX/RxJava#6177]
I recently found that `Completable.takeUntil()` had been added since RxJava 2.1.17.
It calls downstream's `onComplete()` when `other` completes before source completes, but in my opinion that's not reasonable.
Because `onComplete()` of `Completable` is conceptually more like an emission rather than just a termination signal, I think it should call `onError(CancellationException)` instead of `onComplete()`, just like `Single.takeUntil()`.

When I write some logic inside the first callback of `subscribe()`, the logic is not called for other stream types such as `Single` or `Observable` when a stream given to `takeUntil()` terminates, but only `Completable` calls the logic with the same form of code.

```kotlin
single.takeUntil(terminationSignal)
    .subscribe {
        // this is not called when terminationSignal terminates earlier
    }

observable.takeUntil(terminationSignal)
    .subscribe {
        // this is also not called when terminationSignal terminates before any emission
    }

completable.takeUntil(terminationSignal)
    .subscribe {
        // this is called even when terminationSignal terminates earlier
    }
```

This is pretty confusing and looks like not consistent, so I suggest to make `takeUntil()` call `onError(CancellationException)` rather than `onComplete()`.
Please consider this issue. Thank you.

**Comment by akarnokd** [comment]
> This is pretty confusing and looks like not consistent

Please consider that the operators have to work with the specific protocols of the participating reactive types. For example, there is no `onComplete` in `Single` thus a `takeUntil` can't complete a `Single` like it would an `Observable`.

> make takeUntil() call onError(CancellationException)

Nothing is preventing you from getting `onError` called on `terminationSignal`:

```java
CompletableSubject terminationSignal = CompletableSubject.create();

completable.takeUntil(terminationSignal)
    .subscribe(() -> {
        System.out.println("Should not be called");
    }, e -> {
        e.printStackTrace();
    });

terminationSignal.onError(new CancellationException());
```

or, if you have a cold termination signal, compose in the exception:

```java
Completable terminationSignal = ...

completable.takeUntil(
        terminationSignal.andThen(
            Completable.fromAction(() -> { throw new CancellationException(); })
        )
    )
    .subscribe(() -> {
        System.out.println("Should not be called");
    }, e -> {
        e.printStackTrace();
    });
```

**Comment by skgmn** [comment]
I had already been doing that way even before `Completable.takeUntil()` was added (using `ambWith()`), but it had been pretty uncomfortable because that made code longer or required me to write my own extension methods (for kotlin) or transformers (for java).
So I hoped the standard `Completable.takeUntil()` to make it easier when it came out, but it was different than I expected. That's why I made my suggestion, but it's ok to be rejected if you have a firm philosophy for designing APIs.
Thank you for the answer.
